p2panda-rs/src/schema/schema.rs

```diff
@@ -0,0 +1,426 @@
+
+/// Struct for building and representing CDDL groups.
+/// CDDL uses groups to define reuseable data structures
+/// they can be merged into schema or used in Vectors, Tables and Structs
+#[derive(Clone, Debug)]
+pub struct Group {
+    name: String,
+    fields: BTreeMap<String, Field>,
+}
+
+impl Group {
+    /// Create a new CDDL group
+    pub fn new(name: String) -> Self {
+        Self {
+            name,
+            fields: BTreeMap::new(),
+        }
+    }
+
+    /// Add an Field to the group.
+    pub fn add_field(&mut self, key: &str, field_type: Field) {
+        self.fields.insert(key.to_owned(), field_type);
+    }
+}
+
+impl fmt::Display for Group {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let map = &self.fields;
+        write!(f, "( ")?;
+        for (count, value) in map.iter().enumerate() {
+            // For every element except the first, add a comma.
+            if count != 0 {
+                write!(f, ", ")?;
+            }
+            write!(f, "{}: {}", value.0, value.1)?;
+        }
+        write!(f, " )")
+    }
+}
+
+/// SchemaBuilder struct for programatically creating CDDL schemas and valdating MessageFields.
+#[derive(Clone, Debug)]
+pub struct SchemaBuilder {
+    name: String,
+    fields: BTreeMap<String, Field>,
+}
+
+/// Schema struct for creating CDDL schemas, valdating MessageFields and creating messages
+/// following the defined schema.
+#[derive(Clone, Debug)]
+pub struct Schema {
+    schema_hash: Hash,
+    schema_string: String,
+}
+
+impl SchemaBuilder {
+    /// Create a new blank Schema
+    pub fn new(name: String) -> Self {
+        Self {
+            name: name,
+            fields: BTreeMap::new(),
+        }
+    }
+
+    /// Add a field definition to this schema
+    pub fn add_message_field(&mut self, key: String, field_type: Type) -> Result<(), SchemaError> {
+        // Match passed type and map it to our MessageFields type and CDDL types (do we still need the
+        // MessageFields type key when we are using schemas?)
+        let type_string = match field_type {
+            Type::Tstr => "str",
+            Type::Int => "int",
+            Type::Float => "float",
+            Type::Bool => "bool",
+            Type::Relation => "relation",
+        };
+
+        // Create a message field group and add fields
+        let mut message_fields = Group::new(key.to_owned());
+        message_fields.add_field("type", Field::String(type_string.to_owned()));
+        message_fields.add_field("value", Field::Type(field_type));
+
+        // Format message fields group as a struct
+        let message_fields = Field::Struct(message_fields);
+
+        // Insert new message field into Schema fields.
+        // If this Schema was created from a cddl string
+        // `fields` will be None.
+        self.fields.insert(key, message_fields);
+        Ok(())
+    }
+
+    /// Validate a message against this user schema
+    #[cfg(not(target_arch = "wasm32"))]
```

Comment:
Maybe with this version of `cddl`: https://crates.io/crates/cddl/0.9.0-beta.0 we can have wasm support?

Reply:
I'll open a new PR for this when this one is merged :+1: